Attributes

nopbyte opened the pull request:
This PR includes the functionality for creating, updating and deleting attributes for devices and the user currently logged in. 

It uses the version of the agile-sdk currently in the security [PR](https://github.com/Agile-IoT/agile-sdk/pull/27) including PDP, PAP, and Audit calls.

nopbyte committed:
Add attribute-related UIs (set, update, delete also with nested attributes)

Change-Type: minor

nopbyte committed:
fix issues and add missing dependency

nopbyte committed:
remove console log

craigmulligan commented on src/reducers/entities.js:
```diff
@@ -35,7 +71,7 @@ export function devicesDiscover(state = [], action) {
 
 
-export function messages(state=[], action) {
+export function messages(state = [], action) {
```
Thanks :)

craigmulligan commented on src/containers/SecurityDetails.js:
```diff
@@ -0,0 +1,293 @@
+
+  render() {
+    var entity = this.props.entity;
+    if (entity && this.props.entityActionsPolicies) {
```
You can use destructuring here to make it a little cleaner. 

```
const { entity, entityActionsPolicies } = this.props;
if (entity && entityActionsPolicies) {
...
```

craigmulligan commented on src/containers/SecurityDetails.js:
```diff
@@ -0,0 +1,293 @@
+    return (<div>{actionButtons}</div>);
+  }
+
+  dataChanged(id, type, data) {
```
Could you explain a little why you need this? Usually this sort of this is handled by redux. The components should just be rendering state and dispatching actions. Rendering is cheap so why not just set it and let react decide what should update? 

```
this.props.setEntityAttributes({
     id, type, data
});
```

craigmulligan commented on src/containers/SecurityDetails.js:
```diff
@@ -0,0 +1,293 @@
+
+
+
+  renderDetails(meta, id, fieldProperties, parent) {
```
Can you split this up into it's own file/module? `SecurityDetails.js` is way to big.

craigmulligan commented on src/containers/SecurityDetails.js:
```diff
@@ -0,0 +1,293 @@
+  renderDetails(meta, id, fieldProperties, parent) {
+
+    if (meta) {
+      var addAttribute = this.canWrite(parent) ? (<ListItem>{this.getAddAttributeFields(parent)}</ListItem>) : null;
```
Let's use `let` or `const` so we have block scoped variables. It's very rare occasions now where `var` is useful.

craigmulligan commented on src/containers/SecurityDetails.js:
```diff
@@ -0,0 +1,293 @@
+              </ListItem>
+            )
+          } else {
+            return (
```
This is very hard to read, could you split the sub components in this `if/else` into there own components. So it's just 

```
if () {
     <childComponent>
} else {
     <otherChildComponent>
}
```

craigmulligan commented on src/containers/SecurityDetails.js:
```diff
@@ -0,0 +1,293 @@
+          attribute = parent + '.' + k;
+        }
+
+        if (fieldProperties && fieldProperties.hidden && fieldProperties.hidden.indexOf(attribute) === -1) {
```
This double if statement too much for anyone to understand at a glance. 

I'd break it into it's own function that explains what it is like `isEditable()`

or try simplify the logic some other way.

craigmulligan commented on src/containers/SecurityDetails.js:
```diff
@@ -0,0 +1,293 @@
+                                 validate={this.customValidateText}
+                                 text={ this.isPrimitive(meta[k]) ? meta[k] :
+                                   this.renderDetails(meta[k], id, fieldProperties, attribute) }
+                                 change={this.dataChanged.bind(this, id, this.props.entityType.replace('/', ''))}
```
You can use arrow functions to do implicit binding, I think it often makes things easier to read. 

eg 
```
change={() => dataChanged(id, this.props.entityType.replace('/', '')}
```

craigmulligan commented on src/containers/SecurityDetails.js:
```diff
@@ -0,0 +1,293 @@
+                                 paramName={attribute}
+              />
+                {
+                  this.renderAttributeActions(id, this.props.entityType.replace('/', ''), attribute)
```
I'd prefer a component here: 
```
<AttributeActions {..props} />
```

craigmulligan commented on src/containers/SecurityDetails.js:
```diff
@@ -0,0 +1,293 @@
+    event.stopPropagation();
+    if (!this.isPrimitive(value)) {
+      var checkExist = setInterval(function () {
+        if (document.getElementsByClassName('editing').length) {
```
With react there should be very few cases that you need to use actual dom methods like `.document`, you should only be concerned with the virtual dom. 

I'd refactor this by storing a `editing` field in the state object, and keeping track of which field is being edited that way.

craigmulligan commented on src/containers/SecurityDetails.js:
```diff
@@ -0,0 +1,293 @@
+           <FloatingActionButton mini={true}  style={addButtonStyle}
+                         label='Add'
+                         onClick={(event) => {
+                           event.stopPropagation();
```
See my comment below about dom methods, you should use internal state to keep track of things.

craigmulligan commented on src/containers/SecurityDetails.js:
```diff
@@ -0,0 +1,293 @@
+
+const actions = {write: ['delete', 'edit', 'add']};
+
+const weight = function (attr, fieldProperties) {
```
What does the weight determine?

craigmulligan commented on src/containers/Entity.js:
```diff
@@ -0,0 +1,68 @@
+  }
+
+  getEntity() {
+    for (var i in this.props.entityList) { //Get the right entity from the entity list
```
Why not use `entityList.find()`?

craigmulligan commented on src/actions/index.js:
```diff
@@ -163,15 +184,178 @@ export const devicesDelete = (deviceId) => {
+      .then((entity) => {
+        return Promise.resolve(entity);
+      })
+      .catch(err => {
```
This should be `then` right?

nopbyte replied:
I don't see the then missing. I don't know if this is what you meant, but there is an issue. If the entity was in IDM, we return the device info from IDM and not from the device manager... So this needs to be fixed, but otherwise, if the device is not in IDM we try to create it in the catch.

craigmulligan commented on src/actions/index.js:
```diff
@@ -163,15 +184,178 @@ export const devicesDelete = (deviceId) => {
+    dispatch(loading(true))
+    var queryObject = [];
+    var actions_seen = [];
+    for (var act in actions) {
```
Use `.map` here. 

Something like:
```
actions_seen = actions.map((action => (actions.hasOwnProperty(act) && actions_seen.indexOf(act) === -1))
```

craigmulligan commented on src/actions/index.js:
```diff
@@ -1,11 +1,30 @@
+//This sets the token for the calls to the sdk and reloads the SDK object
+
+export const setToken = (new_token) => {
+  var token = new_token;
```
Once https://github.com/Agile-IoT/agile-sdk/pull/30 is set we can just use `agile.setToken`

craigmulligan commented on src/actions/index.js:
```diff
@@ -163,15 +184,178 @@ export const devicesDelete = (deviceId) => {
+      results => {
+        var result_list = {id: id, type: type, policies: {}};
+        var i = 0;
+        attribute_names.forEach(attribute => {
```
Not exactly sure what these loops are doing, but perhaps check if there is a lodash function to simplify the logic?

nopbyte replied:
This is an optimization that we had to do, to query less times the policies. Otherwise we end up querying the same thing too many times. This could have performance impact on the PDP, considering that it only uses a levelDB database. So, I am afraid there is no way around this, if we want to keep an eye on resource consumption on the gateway's side.

nopbyte replied:
This is an optimization that we had to do, to query less times the policies. Otherwise we end up querying the same thing too many times. This could have performance impact on the PDP, considering that it only uses a levelDB database. So, I am afraid there is no way around this, if we want to keep an eye on resource consumption on the gateway's side.

craigmulligan reviewed (changes requested):
Looks like you're grasping react pretty quickly :). my main concern at the moment is readability, there were a few sections where I really couldn't follow the code, mainly in `components`. Let me know if you have any questions regarding the comments.

nopbyte committed:
address comments for readability and use of state provided during the code review.

nopbyte commented:
Hi @craig-mulligan, 

Can you please have a look again?

We have addressed your comments. In the end, we modified the way in which non primitive (objects) attributes were set. This was introducing many of the issues you pointed out, because we needed to handle events from the edditable field in a way that was not intended, to be able to parse and stringify json objects. Now, we just used buttons and by creating an attribute with name, but without value the UI creates an empty object underneath... We also used a new library to handle some aspects of traversing the objects and we tried to balance better what the components and containers do to improve readability in general. I hope it looks better now!

This version includes the agile-sdk in its version including the tokenSet function and it is also updated in the code. So it should be ready to be included, as long as the osjs version is using the latest that includes the token when the UI Iframe is opened in OSJS.

Thanks for your review

craigmulligan reviewed (approved):
Nice! 👏
